fendermint/vm/interpreter/src/fvm/state/ipc.rs

```diff
@@ -270,10 +258,26 @@ impl<DB: Blockstore> GatewayCaller<DB> {
     }
 }
 
-/// Hash some value in the same way we'd hash it in Solidity.
+/// Total amount of tokens to mint as a result of top-down messages arriving at the subnet.
 ///
-/// Be careful that if we have to hash a single struct, Solidity's `abi.encode`
-/// function will treat it as a tuple.
-pub fn abi_hash<T: Tokenize>(value: T) -> [u8; 32] {
-    keccak256(ethers::abi::encode(&value.into_tokens()))
+/// Fees are ignored, which suggests that they will be paid similarly to gas,
+/// from the balance of the sender already present in the currents subnet.
+pub fn tokens_to_mint(msgs: &[ipc_sdk::cross::CrossMsg]) -> TokenAmount {
+    msgs.iter()
+        .fold(TokenAmount::from_atto(0), |mut total, msg| {
+            total += &msg.msg.value;
+            total
+        })
+}
+
+/// Total amount of tokens to burn as a result of bottom-up messages leaving the subnet.
+///
+/// Fees are ignored, which suggests that they will be paid similarly to gas,
+/// from the balance of the sender already present in the parent subnet.
+pub fn tokens_to_burn(msgs: &[gateway_getter_facet::CrossMsg]) -> TokenAmount {
+    msgs.iter()
+        .fold(TokenAmount::from_atto(0), |mut total, msg| {
+            total += from_eth::to_fvm_tokens(&msg.message.value);
```

Comment:
Same here, we should reduce from the circulating supply the fees and the value of the cross-net message (you could also look at the `value` field for the wrapping release message, but this is not the case as you are directly looking at what is being propagated, but for context: https://github.com/consensus-shipyard/ipc-solidity-actors/blob/dev/src/gateway/GatewayManagerFacet.sol#L143-L150)

Reply:
Great link, thanks!